### PR TITLE
Add RAG logic to AppEngine app scaffold

### PR DESCRIPTION
**Pull request by BenoitDherin**
This PR adds the RAG logic to the AppEngine scaffolds. 

The main files to review are:

1. server.py
2. rag.py
3. services.py
4. settings.py

Proceeding in order will help understand the logics.

The testing instructions are in the README, which contains new targets: `make setup` to provision and authenticate the necessary services and `make ingest` to ingest the data into the vector store. These targets need to be run before any testing. 

**Note:** `make notebook` will launch a jupyterlab for local developement and debugging with the app libs already loaded by default (i.e. one can do `from app import lib`)

### CODE REVIEW

**Review by sanjanalreddy — APPROVED** [review]
The app deployed as expected. Thank you for adding this!

File: scaffolds/langchain_rag/app/rag.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+""" Application logic: Assemble data from all services.
+"""
+
+from app.services import create_rag_service
+
+_NO_PROMPT = "Please enter a question."
+_SUFFIX = """
```
[the remark this answers is not in the document]
Comment (BenoitDherin) [reply scaffolds/langchain_rag/app/rag.py:7]: Done!

File: scaffolds/langchain_rag/app/rag.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+def create_rag():
+    rag_svc = create_rag_service()
+    return Rag(rag_svc)
+
+
+def _extract_link_from_source(source):
+    url = source.metadata.get("url")
+    html = None
+    if url:
```
[the remark this answers is not in the document]
Comment (BenoitDherin) [reply scaffolds/langchain_rag/app/rag.py:46]: Great catch! I added some doc at the beginning of the file for now, and will remove that dependency in a follow up PR.

File: scaffolds/langchain_rag/app/services.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+import os
+
+from app import lib
+from app.settings import Config
+from google.cloud import aiplatform
+from langchain.chains import RetrievalQA
+from langchain.document_loaders import DataFrameLoader
+from langchain.embeddings import VertexAIEmbeddings
+from langchain.llms import VertexAI
```
[the remark this answers is not in the document]
Comment (BenoitDherin) [reply scaffolds/langchain_rag/app/services.py:12]: As discussed, update to Gemini will be done in a follow up PR.

File: scaffolds/langchain_rag/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+```bash
+make setup
+```
+
+The next step is to create and populate the virtual environment with
+
+```bash
+make venv
+```
+After this step you should find a new folder called `venv` containing the virtual environment.
```
[the remark this answers is not in the document]
Comment (BenoitDherin) [reply scaffolds/langchain_rag/README.md:41]: Done!

**Pull request merged**